<a href="https://colab.research.google.com/github/christophermalone/DSCI325/blob/main/Module3_Part3_R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 3 | Part 3 | R : MUTATE()

This notebook will discuss the use of the MUTATE() data verb in R.

<table width='100%' ><tr><td bgcolor='green'></td></tr></table>

### Example 3.3.1R
For this notebook, consider the Nursing Home Provider dataset.  The datasets is published by the Centers for Medicare and Medicade Services.

<table>
<tr>
<td colspan="2" align='center'>Fields in Dataset</td>
</tr>
<tr>
<td valign='top' align='left'>
1: FPN (Federal Provider Number)<br>
2: ProviderName<br>
3: ProviderAddress<br>
4: ProviderCounty<br>
5: OwnershipType<br><br>
6: NumberBeds<br>
7: AverageResidentsperDay<br><br>
8: ProviderInHospital<br>
9: BusinessName<br>
10: ContinuingCareRetirementCommunity<br><br>
11: HealthInspection_MoreThan2YearsAgo<br>
12: Overall_Rating<br>
13: HealthInspection_Rating<br>
14: QM_Rating<br>
15: LongStayQM_Rating<br>
16: ShortStayQM_Rating<br>
17: Staffing_Rating<br>
18: RNStaffing_Rating<br>
</td>
<td valign='top' align='left'>
19: NurseAideStaffing_Hours<br>
20: LPNStaffing_Hours<br>
21: RNStaffing_Hours<br>
22: LicensedStaffing_Hours<br>
23: TotalNurseStaffing_Hours<br>
24: PhysicalTherapistStaffing_Hours<br><br>
25: RatingCycle1_Date<br>
26: RatingCycle1_HealthDeficiencyScore<br>
27: RatingCycle2_Date<br>
28: RatingCycle2_HealthDeficiencyScore<br>
29: RatingCycle3_Date<br>
30: RatingCycle3_HealthDeficiencyScore<br><br>
31: NumberofFines<br>
32: TotalAmountFines<br>
</td>
</tr>
</table>

 
<br>Original Data Source: https://data.cms.gov/provider-data/search?theme=Nursing%20homes%20including%20rehab%20services

<table width='100%' ><tr><td bgcolor='green'></td></tr></table>

Load the tidyverse() package.

In [4]:
library(tidyverse)

Reading in the ProviderInfo.csv dataset using read_csv().

In [5]:
NursingHomes <- read_csv("/content/sample_data/ProviderInfo.csv")


── Column specification ────────────────────────────────────────────────────────
cols(
  .default = col_double(),
  FPN = col_character(),
  ProviderName = col_character(),
  ProviderAddress = col_character(),
  ProviderCounty = col_character(),
  OwnershipType = col_character(),
  ProviderInHospital = col_character(),
  BusinessName = col_character(),
  ContinuingCareRetirementCommunity = col_character(),
  HealthInspection_MoreThan2YearsAgo = col_character(),
  RatingCycle1_Date = col_character(),
  RatingCycle1_HealthDeficiencyScore = col_character(),
  RatingCycle2_Date = col_character(),
  RatingCycle2_HealthDeficiencyScore = col_character(),
  RatingCycle3_Date = col_character(),
  RatingCycle3_HealthDeficiencyScore = col_character()
)
ℹ Use `spec()` for the full column specifications.




In [6]:
#Getting the dimension
cat("The number of rows and columns in this dataframe is: \n")
dim(NursingHomes)
cat("\n")


#Getting a list of all the field names
cat("Field names in the Nursing Homes Dataset: \n")
colnames(NursingHomes)

#Seeing first few rows
head(NursingHomes)


The number of rows and columns in this dataframe is: 


[1] 15331    32


Field names in the Nursing Homes Dataset: 


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"

FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ContinuingCareRetirementCommunity ⋯
1 BURNS NURSING HOME, INC.             N                                 ⋯
2 COOSA VALLEY HEALTHCARE CENTER LLC   N                                 ⋯
3 JACKSON COUNTY HEALTH CARE AUTHORITY N                                 ⋯
4 BALL HEALTHCARE EASTVIEW INC         N                                 ⋯
5 C & G HEALTHCARE SERVICES, INC.      N                                 ⋯
6 ATHENS HEALTH AND REHABILITATION LLC N                                 ⋯
  TotalNurseStaffing_Hours PhysicalTherapistStaffing_Hours RatingCycle1_Date
1 4.31259                  0.00000                         8/21/2019        
2 4.84021                  0.01271                         6/13/2019        
3 4.76645                  0.05596                         6/6/2019         
4 3.81920                  0.03246                         2/20/2020        
5 3.39949                  0.01248                         3/5/2020         
6 3.24110                  0.03716                         8/29/2019        
  RatingCycle1_HealthDeficiencyScore RatingCycle2_Date
1 8                                  8/1/2018         
2 4                                  6/7/2018         
3 20                                 5/3/2018         
4 4                                  1/24/2019        
5 20                                 5/2/2019         
6 20                                 7/19/2018        
  RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 4                                  6/22/2017        
2 32                                 4/6/2017         
3 40                                 3/16/2017        
4 20                                 1/25/2018        
5 4                                  4/19/2018        
6 28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines
1 0                                  0                 0           
2 36                                 1               650           
3 44                                 0                 0           
4 24                                 0                 0           
5 28                                 2             29611           
6 20                                 0                 0

# Create New Field - Numeric

Nursing homes are most profitable when they have lots of residents, i.e. are near capacity.  Consider the following new field to be calculated for this data.

$$Capacity = \frac{Average \space Residents \space per \space Day}{Number \space of \space Beds}$$

The <strong>mutate()</strong> function in the <strong>dplyr</strong> package will can be used to complete this task.  The mutate() function has the following syntax.

In [7]:
(
  NursingHomes
  %>% mutate(Capacity = AverageResidentsperDay / NumberBeds)
  %>% head()
) 

FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ContinuingCareRetirementCommunity ⋯
1 BURNS NURSING HOME, INC.             N                                 ⋯
2 COOSA VALLEY HEALTHCARE CENTER LLC   N                                 ⋯
3 JACKSON COUNTY HEALTH CARE AUTHORITY N                                 ⋯
4 BALL HEALTHCARE EASTVIEW INC         N                                 ⋯
5 C & G HEALTHCARE SERVICES, INC.      N                                 ⋯
6 ATHENS HEALTH AND REHABILITATION LLC N                                 ⋯
  PhysicalTherapistStaffing_Hours RatingCycle1_Date
1 0.00000                         8/21/2019        
2 0.01271                         6/13/2019        
3 0.05596                         6/6/2019         
4 0.03246                         2/20/2020        
5 0.01248                         3/5/2020         
6 0.03716                         8/29/2019        
  RatingCycle1_HealthDeficiencyScore RatingCycle2_Date
1 8                                  8/1/2018         
2 4                                  6/7/2018         
3 20                                 5/3/2018         
4 4                                  1/24/2019        
5 20                                 5/2/2019         
6 20                                 7/19/2018        
  RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 4                                  6/22/2017        
2 32                                 4/6/2017         
3 40                                 3/16/2017        
4 20                                 1/25/2018        
5 4                                  4/19/2018        
6 28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines Capacity 
1 0                                  0                 0            0.9087719
2 36                                 1               650            0.8482353
3 44                                 0                 0            0.8260000
4 24                                 0                 0            0.7804348
5 28                                 2             29611            0.8135922
6 20                                 0                 0            0.8402685

<u>Question:</u>  Did this new field automatically get added to the NursingHomes dataframe (or did this just get printed to the screen)?

<u>Answer:</u> Check the dimension of the dataframe.  The existing dataframe had 15331 rows x 32 columns.

In [8]:
#What is the dimension of the NursingHomes dataframe?
dim(NursingHomes)

cat("\n")
cat("Field names: \n")
colnames(NursingHomes)

[1] 15331    32


Field names: 


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"

In the following snipit of code, the outcome for the sequnce of commands is pushed into the NursingHome dataframe.  This idea is equivalent to "edit-in-place" option in SED.

In [9]:
(
  NursingHomes
  %>% mutate(Capacity = AverageResidentsperDay / NumberBeds)
  #%>% head()
) -> NursingHomes 

In [10]:
#What is the dimension of the NursingHomes dataframe now?
dim(NursingHomes)

cat("\n")
cat("Field names: \n")
colnames(NursingHomes)

[1] 15331    33


Field names: 


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"

<u>Comment</u>: The Capacity quantity is not defined when the denominitor is zero.  Thus, before creating this new field, one should verify that all <i>Number of Beds</i> values are greater than 0. 

A more sound approach, would be to use the following definition of Capacity.


$$Capacity = \Bigg\{ \begin{array} \tiny{\frac{\mbox{Average Residents per Day}}{\mbox{Number of Beds}}} & when & \mbox{Number of Beds} > 0\\
NA &  when & \mbox{Number of Beds} = 0
\end{array}$$



In [11]:
(
  NursingHomes
  %>% mutate(Capacity = ifelse(NumberBeds > 0, AverageResidentsperDay / NumberBeds, NA) )
) -> NursingHomes 


Taking a look at the output produced from the code above.

In [12]:
#What is the dimension of the NursingHomes dataframe now?
dim(NursingHomes)

cat("\n")
cat("Field names: \n")
colnames(NursingHomes)

cat("First few rows: \n")
head(NursingHomes)


[1] 15331    33


Field names: 


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"

First few rows: 


FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ContinuingCareRetirementCommunity ⋯
1 BURNS NURSING HOME, INC.             N                                 ⋯
2 COOSA VALLEY HEALTHCARE CENTER LLC   N                                 ⋯
3 JACKSON COUNTY HEALTH CARE AUTHORITY N                                 ⋯
4 BALL HEALTHCARE EASTVIEW INC         N                                 ⋯
5 C & G HEALTHCARE SERVICES, INC.      N                                 ⋯
6 ATHENS HEALTH AND REHABILITATION LLC N                                 ⋯
  PhysicalTherapistStaffing_Hours RatingCycle1_Date
1 0.00000                         8/21/2019        
2 0.01271                         6/13/2019        
3 0.05596                         6/6/2019         
4 0.03246                         2/20/2020        
5 0.01248                         3/5/2020         
6 0.03716                         8/29/2019        
  RatingCycle1_HealthDeficiencyScore RatingCycle2_Date
1 8                                  8/1/2018         
2 4                                  6/7/2018         
3 20                                 5/3/2018         
4 4                                  1/24/2019        
5 20                                 5/2/2019         
6 20                                 7/19/2018        
  RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 4                                  6/22/2017        
2 32                                 4/6/2017         
3 40                                 3/16/2017        
4 20                                 1/25/2018        
5 4                                  4/19/2018        
6 28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines Capacity 
1 0                                  0                 0            0.9087719
2 36                                 1               650            0.8482353
3 44                                 0                 0            0.8260000
4 24                                 0                 0            0.7804348
5 28                                 2             29611            0.8135922
6 20                                 0                 0            0.8402685



---

---





**Task #1**:  Consider the following code -- answer the questions that follow. 


In [13]:
(
  NursingHomes
  %>% mutate(Capacity = ifelse(NumberBeds > 0, AverageResidentsperDay / NumberBeds, NA) )
  %>% select(FPN, ProviderName, Capacity)
  %>% arrange(desc(Capacity))
  %>% filter(Capacity > 1)
  %>% dim()
)

[1] 100   3


1.   What is the purpose of Line 4? Discusss

2.   What is the purpose of Line 5? Discuss

3.   What is is the purpose of Line 6? Discuss

4.   What information is gained by running this code?  Discuss



---

---





**Task #2**:  Consider the following code -- answer the questions that follow. 

In [14]:
(
  NursingHomes
  %>% mutate(Capacity = ifelse(NumberBeds > 0, 100*round(AverageResidentsperDay / NumberBeds,3), NA) )
  %>% select(FPN, ProviderName, Capacity)
  %>% arrange(desc(Capacity))
  %>% filter(Capacity <= 100)
  %>% mutate(Capacity = as.character(Capacity))
  %>% mutate(Capacity = paste(Capacity, "%", sep=""))
  %>% head(10)
)


FPN    ProviderName                                      Capacity
1  35287  HAVASU REGIONAL MEDICAL CENTER                    100%    
2  05A192 THE ROYAL HOME                                    100%    
3  16E668 PALO ALTO COUNTY HOSPITAL                         100%    
4  335874 THE STEVEN AND ALEXANDRA COHEN PED L T C PAVILION 100%    
5  495405 SUMMIT SQUARE                                     100%    
6  515182 MILETREE CENTER                                   100%    
7  555450 AMERICAN RIVER CENTER                             100%    
8  105138 HEALTHCARE CENTER OF JACKSONVILLE                 99.8%   
9  675509 PARKLANE WEST HEALTHCARE CENTER                   99.8%   
10 235624 PORTAGEPOINTE                                     99.7%

1.   Change this code to output 20 nursing homes instead of the list provided here.

2.   What would happen if Line 7 were placed before Line 6?  How would this change effect Line 6?

3.   Figure out how to save a csv file of this output. Put this csv file into the '/content/sample_data/' Colab folder.



---

---







Next, calculate the following quantity for each nursing home.

$$ Average \space Fine = \frac{\mbox{Total Amount of Fines}}{\mbox{Number of Fines} }$$

Before, proceeding one should check that this quantity can be computed for all nursing homes. If the Number of Fines is 0, then this quantity is not defined.

In [15]:
#Getting a summary of all fields in R
summary(NursingHomes)


     FPN            ProviderName       ProviderAddress    ProviderCounty    
 Length:15331       Length:15331       Length:15331       Length:15331      
 Class :character   Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character  
                                                                            
                                                                            
                                                                            
                                                                            
 OwnershipType        NumberBeds     AverageResidentsperDay ProviderInHospital
 Length:15331       Min.   :   2.0   Min.   :  1.00         Length:15331      
 Class :character   1st Qu.:  65.0   1st Qu.: 45.70         Class :character  
 Mode  :character   Median :  99.0   Median : 68.60         Mode  :character  
                    Mean   : 106.5   Mean   : 75.29                 

In [16]:
#Getting the summary for Number of Fines only
summary(NursingHomes$NumberofFines)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.0000  0.0000  0.0000  0.4342  1.0000  9.0000 

Given that many of the nursing homes do not have any fines, the followign should be used for the calcuation of average fine amount.

$$AverageFine = \Bigg\{ \begin{array} \tiny{\frac{\mbox{Total Amount of Fines}}{\mbox{Number of Fines}}} & when & \mbox{Number of Fines} > 0\\
0 &  when & \mbox{Number of Fines} = 0
\end{array}$$

In [17]:
(
  NursingHomes
  %>% mutate(AvgFine = ifelse(NumberofFines > 0, TotalAmountFines / NumberofFines, 0) )
) -> NursingHomes

#Making sure column was added
dim(NursingHomes)
cat("\n")
colnames(NursingHomes)


[1] 15331    34

[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"                           "AvgFine"

Next, consider the average (and median) for all nursing homes.

In [18]:
(
  NursingHomes
  %>% summarize('Average of Fines Overall' = mean(AvgFine), 'Median of Fines Overall' = median(AvgFine))
)

Average of Fines Overall Median of Fines Overall
1 8334.139                 0

Next, compute these quantites for Nursing Home that had a non-zero fine.

In [19]:
(
  NursingHomes
  %>% filter(AvgFine > 0)
  %>% summarize('Average of Fines Overall' = mean(AvgFine), 'Median of Fines Overall' = median(AvgFine))
)

Average of Fines Overall Median of Fines Overall
1 27459.85                 11950

# Create New Field - Character

Next, suppose the desire is to get the average Overall_Rating for nursing homes in MN.  Notice that State is not currently a field.

In [20]:
colnames(NursingHomes)

[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"                           "AvgFine"

Realize that the State can be pulled off from the ProviderAddress or Provider County.  

In [21]:
(
  NursingHomes
  %>% select(ProviderAddress, ProviderCounty)
  %>% head()
)

ProviderAddress                               ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        Limestone County, AL

To pull off elements of a string, we will use the <strong>str_sub</strong> function from the <strong>stringr</strong> pakcage. The stringr package is part of the <strong>tidyverse</strong> package.

<p align='center'><img src="https://drive.google.com/uc?export=view&id=1gxaqqR7oxvPYQJYDiJDIMnwce_I_by_o" width='75%' height='75%'></p>


The syntax for the <strong>str_sub()</strong> function is


<p align='center'>
<font size='+2'>str_sub(<i>field name</i>, <i>start =</i> , <i>end =</i> )</font>
</p>

Note:  Negative indicies for <i>start</i> and <i>end</i> can be used to search a string <strong>backwards</strong>.


In [22]:
(
 NursingHomes
 %>% mutate(State = str_sub(ProviderCounty, -2,-1) )
) -> NursingHomes

colnames(NursingHomes)


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"                           "AvgFine"                           
[35] "State"

Notice that R always appears to add the newly created column from mutate() to the far right side of the dataframe.  There exists a <strong>relocate()</strong> function that can be used to rearrange the order of the columns.

In [23]:
(
 NursingHomes
 %>% relocate(State, .after = ProviderAddress )
) -> NursingHomes

colnames(NursingHomes)
cat("\n")
head(NursingHomes)

[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "State"                             
 [5] "ProviderCounty"                     "OwnershipType"                     
 [7] "NumberBeds"                         "AverageResidentsperDay"            
 [9] "ProviderInHospital"                 "BusinessName"                      
[11] "ContinuingCareRetirementCommunity"  "HealthInspection_MoreThan2YearsAgo"
[13] "Overall_Rating"                     "HealthInspection_Rating"           
[15] "QM_Rating"                          "LongStayQM_Rating"                 
[17] "ShortStayQM_Rating"                 "Staffing_Rating"                   
[19] "RNStaffing_Rating"                  "NurseAideStaffing_Hours"           
[21] "LPNStaffing_Hours"                  "RNStaffing_Hours"                  
[23] "LicensedStaffing_Hours"             "TotalNurseStaffing_Hours"          
[25] "PhysicalTherapistStaffing_Hours"    "RatingCycle1_Date"                 
[27] "RatingCycle1_HealthDeficiencyScore" "RatingCycle2_Date"                 
[29] "RatingCycle2_HealthDeficiencyScore" "RatingCycle3_Date"                 
[31] "RatingCycle3_HealthDeficiencyScore" "NumberofFines"                     
[33] "TotalAmountFines"                   "Capacity"                          
[35] "AvgFine"

FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               State ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    AL    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     AL    Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       AL    Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  AL    Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 AL    Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        AL    Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ⋯ RatingCycle1_Date
1 BURNS NURSING HOME, INC.             ⋯ 8/21/2019        
2 COOSA VALLEY HEALTHCARE CENTER LLC   ⋯ 6/13/2019        
3 JACKSON COUNTY HEALTH CARE AUTHORITY ⋯ 6/6/2019         
4 BALL HEALTHCARE EASTVIEW INC         ⋯ 2/20/2020        
5 C & G HEALTHCARE SERVICES, INC.      ⋯ 3/5/2020         
6 ATHENS HEALTH AND REHABILITATION LLC ⋯ 8/29/2019        
  RatingCycle1_HealthDeficiencyScore RatingCycle2_Date
1 8                                  8/1/2018         
2 4                                  6/7/2018         
3 20                                 5/3/2018         
4 4                                  1/24/2019        
5 20                                 5/2/2019         
6 20                                 7/19/2018        
  RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 4                                  6/22/2017        
2 32                                 4/6/2017         
3 40                                 3/16/2017        
4 20                                 1/25/2018        
5 4                                  4/19/2018        
6 28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines Capacity 
1 0                                  0                 0            0.9087719
2 36                                 1               650            0.8482353
3 44                                 0                 0            0.8260000
4 24                                 0                 0            0.7804348
5 28                                 2             29611            0.8135922
6 20                                 0                 0            0.8402685
  AvgFine
1     0.0
2   650.0
3     0.0
4     0.0
5 14805.5
6     0.0

Next, consider the code that provides a summary of the total amount of fines across the entrie state of MN and WI.  The <strong>message()</strong> function is unique to Colab and forces printing to the screen.

In [24]:
(
 NursingHomes
 %>% filter(State == "MN" | State == "WI")
 %>% group_by(State)
 %>% summarize(TotalFines = sum(TotalAmountFines))
 ) -> MNandWI_Fines

MNandWI_Fines

#message("The total amount of fines for nursing homes in MN is ", MNandWI_Fines[1,2],".")
#message("The total amount of fines for nursing homes is WI is ", MNandWI_Fines[2,2],".")




State TotalFines
1 MN    2033069   
2 WI    3305504

Consider the <strong>dollar()</strong> function in the <strong>scales</strong> package.  The dollar() function will improve formatting in this situation.  

In [25]:
library(scales)


Attaching package: ‘scales’


The following object is masked from ‘package:purrr’:

    discard


The following object is masked from ‘package:readr’:

    col_factor




The dollar() function must take a numeric data type.  In this situation, to remove the data frame structure, use either <strong>unlist()</strong> or <strong>as.numeric</strong>.  Each is used here.

In [26]:
(
 NursingHomes
 %>% filter(State == "MN" | State == "WI")
 %>% group_by(State)
 %>% summarize(TotalFines = sum(TotalAmountFines))
 ) -> MNandWI_Fines

message("The total amount of fines for nursing homes in MN is ", dollar(unlist(MNandWI_Fines[1,2])),".")
message("The total amount of fines for nursing homes is WI is ", dollar(as.numeric(MNandWI_Fines[2,2])),".")



The total amount of fines for nursing homes in MN is $2,033,069.

The total amount of fines for nursing homes is WI is $3,305,504.



Next, let us consider summaries based on OwnershipType.  OwnershipType is a field that is compound in nature, i.e. contains two different bits of information seperated by a dash.    To pull this informaton apart, the <strong>str_split()</strong> will be used inside the mutate() function.


In [27]:
(
 NursingHomes
 %>% select(OwnershipType)
 %>% head()
)

OwnershipType           
1 For profit - Corporation
2 For profit - Corporation
3 Government - County     
4 For profit - Individual 
5 For profit - Individual 
6 For profit - Corporation

The syntax for the <strong>str_split()</strong> function is provided here.

<p align='center'>
<font size='+2'>str_split(<i>field name</i>, <i>fixed( character )</i> , <i>n =   </i> )</font>
</p>


The output produced by the str_split() function is a list.  This list is contained <strong>within</strong> the dataframe. 

<p align='center'><img src="https://drive.google.com/uc?export=view&id=1ar07bxYXT2IbXD-Or-2gWyf6nTrDMT8S" width='75%' height='75%'></p>

In [28]:
(
 NursingHomes
 %>% mutate(OwnershipType_Split = str_split(OwnershipType, fixed("-"),n=2))
) -> NursingHomes_OwnershipSplit

head(NursingHomes_OwnershipSplit)

FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               State ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    AL    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     AL    Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       AL    Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  AL    Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 AL    Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        AL    Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ⋯ RatingCycle1_HealthDeficiencyScore
1 BURNS NURSING HOME, INC.             ⋯ 8                                 
2 COOSA VALLEY HEALTHCARE CENTER LLC   ⋯ 4                                 
3 JACKSON COUNTY HEALTH CARE AUTHORITY ⋯ 20                                
4 BALL HEALTHCARE EASTVIEW INC         ⋯ 4                                 
5 C & G HEALTHCARE SERVICES, INC.      ⋯ 20                                
6 ATHENS HEALTH AND REHABILITATION LLC ⋯ 20                                
  RatingCycle2_Date RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 8/1/2018          4                                  6/22/2017        
2 6/7/2018          32                                 4/6/2017         
3 5/3/2018          40                                 3/16/2017        
4 1/24/2019         20                                 1/25/2018        
5 5/2/2019          4                                  4/19/2018        
6 7/19/2018         28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines Capacity 
1 0                                  0                 0            0.9087719
2 36                                 1               650            0.8482353
3 44                                 0                 0            0.8260000
4 24                                 0                 0            0.7804348
5 28                                 2             29611            0.8135922
6 20                                 0                 0            0.8402685
  AvgFine OwnershipType_Split       
1     0.0 For profit  ,  Corporation
2   650.0 For profit  ,  Corporation
3     0.0 Government ,  County      
4     0.0 For profit ,  Individual  
5 14805.5 For profit ,  Individual  
6     0.0 For profit  ,  Corporation

The following code will split the list created by str_split() into separate fields in the dataframe.

<p align='center'><img src="https://drive.google.com/uc?export=view&id=1OHRxCGKU0AByLB-uum9iq6jgy4bbHP9p" width='75%' height='75%'></p>

In [33]:
(
 NursingHomes
 %>% mutate(OwnershipType_Split = str_split(OwnershipType, fixed("-"),n=2))
 %>% mutate(ProfitType = map_chr(OwnershipType_Split, 1), CorporationType = map_chr(OwnershipType_Split,2))
 ) -> NursingHomes_OwnershipSplit


Next, identify the levels for ProfitType field.

In [34]:
(
 NursingHomes_OwnershipSplit
 %>% filter(State == "MN")
 %>% group_by(ProfitType)
 %>% summarize(TotalFines = sum(TotalAmountFines), Count = n())
 %>% mutate('Average Fine Amount' = dollar(TotalFines/Count))
) 

ProfitType  TotalFines Count Average Fine Amount
1 For profit   740538    114   $6,495.95          
2 Government    90912     31   $2,932.65          
3 Non profit  1201619    222   $5,412.70



---

---







Consider the <strong>case_when()</strong> function.  This is an extension to the <strong>ifelse()</strong> statement that allows for more than a single condition to be checked. It is true that ifelse() statements can be nested to allow for the checking of multiple conditions, but most encourage the use of the case_when() function in this situation.


<table>
<tr>
<td align="center">ProfitType</td>
<td align="center">For Profit</td>
</tr>
<tr>
 <td align="center" bgcolor="black"><font size="+1" color="white">For profit</font></td>
 <td align="center"bgcolor="black"><font size="+1" color="white">Yes</font></td>
</tr>
<tr>
 <td align="center" bgcolor="lightgray"><font size="+1">Government</font></td>
 <td align="center" rowspan="2" bgcolor="lightgray"><font size="+1">No</font></td>
</tr>
<tr>
 <td align="center" bgcolor="lightgray"><font size="+1">Non profit</font></td>
</td>
</table>



In [54]:
(
 NursingHomes_OwnershipSplit
  %>% mutate(ForProfit = 
                          case_when(
                                    ProfitType == "For profit" ~ "Yes",
                                    ProfitType == "Government" ~ "No",
                                    ProfitType == "Non profit" ~ "No"
                          )
            )
   %>% select(OwnershipType_Split, ProfitType, CorporationType, ForProfit)
 ) -> temp

head(temp)

OwnershipType_Split        ProfitType  CorporationType ForProfit
1 For profit  ,  Corporation For profit   Corporation    NA       
2 For profit  ,  Corporation For profit   Corporation    NA       
3 Government ,  County       Government   County         NA       
4 For profit ,  Individual   For profit   Individual     NA       
5 For profit ,  Individual   For profit   Individual     NA       
6 For profit  ,  Corporation For profit   Corporation    NA

Unfornately, this does not work as the ForProfit column contains all NA values.

<strong>Question</strong>: Why is this happening?<br>
<strong>Answer</strong>: Most likely the issue is with the condition being checked.


Let us consider the pantom spaces that are often present in strings!  Consider the following code that computes the length of each part of the string split above.  

Notice that the string "For profit" has 10 characters, but the the string length function says there are 11 character present.  Thus, an additional character exists within the "For profit" string -- likely the actual string is "For profit_".  Regarding CorporationType, the actual string is likely "_Corporation" as this is the right side of the string split done above.

In [52]:
(
 NursingHomes_OwnershipSplit
 %>% select(OwnershipType_Split, ProfitType, CorporationType)
 %>% mutate(ProfitTypeLen = str_length(ProfitType), CorporationTypeLen = str_length(CorporationType))
 ) -> temp

head(temp)


OwnershipType_Split        ProfitType  CorporationType ProfitTypeLen
1 For profit  ,  Corporation For profit   Corporation    11           
2 For profit  ,  Corporation For profit   Corporation    11           
3 Government ,  County       Government   County         11           
4 For profit ,  Individual   For profit   Individual     11           
5 For profit ,  Individual   For profit   Individual     11           
6 For profit  ,  Corporation For profit   Corporation    11           
  CorporationTypeLen
1 12                
2 12                
3  7                
4 11                
5 11                
6 12

The <strong>str_trim()</strong> function, called trim() in most languages, can be used to remove all empty spaces at the beginning and end of a string.  The trim() function typically does not remove spaces in the middle of a string.

A alternative would be to use the str_replace() function.  The str_replace() could be used to replace all spaces with nothing; however, str_replace() would remove <strong>all</strong> spaces -- including the space between "For" and "profit" in the string "For profit", e.g. "For profit" would become "Forprofit" with str_replace().

In [62]:
(
 NursingHomes_OwnershipSplit
  %>% mutate(ForProfit = 
                          case_when(
                                    str_trim(ProfitType) == "For profit" ~ "Yes",
                                    str_trim(ProfitType) == "Government" ~ "No",
                                    str_trim(ProfitType) == "Non profit" ~ "No"
                          )
  
            )
   %>% select(OwnershipType_Split, ProfitType, CorporationType, ForProfit, State)
 ) -> temp

head(temp, 20)

#The ForProfit variable can now be used, e.g. count the number of For Profit and Non Profits in MN
cat("\n")
cat("\n")
cat("\n")
cat("Number of For Profit and and Non Profit Nursing Homes in MN.\n")
cat("\n")
(
 temp
 %>% filter(State == "MN")
 %>% group_by(ForProfit)
 %>% summarize(Count=n())
)

OwnershipType_Split                                    ProfitType 
1  For profit  ,  Corporation                             For profit 
2  For profit  ,  Corporation                             For profit 
3  Government ,  County                                   Government 
4  For profit ,  Individual                               For profit 
5  For profit ,  Individual                               For profit 
6  For profit  ,  Corporation                             For profit 
7  For profit  ,  Corporation                             For profit 
8  For profit  ,  Corporation                             For profit 
9  For profit  ,  Corporation                             For profit 
10 For profit  ,  Corporation                             For profit 
11 For profit  ,  Corporation                             For profit 
12 For profit  ,  Corporation                             For profit 
13 For profit  ,  Corporation                             For profit 
14 For profit  ,  Corporation                             For profit 
15 For profit                ,  Limited Liability company For profit 
16 For profit  ,  Corporation                             For profit 
17 For profit  ,  Corporation                             For profit 
18 For profit  ,  Corporation                             For profit 
19 For profit  ,  Corporation                             For profit 
20 Non profit ,  Other                                    Non profit 
   CorporationType            ForProfit State
1   Corporation               Yes       AL   
2   Corporation               Yes       AL   
3   County                    No        AL   
4   Individual                Yes       AL   
5   Individual                Yes       AL   
6   Corporation               Yes       AL   
7   Corporation               Yes       AL   
8   Corporation               Yes       AL   
9   Corporation               Yes       AL   
10  Corporation               Yes       AL   
11  Corporation               Yes       AL   
12  Corporation               Yes       AL   
13  Corporation               Yes       AL   
14  Corporation               Yes       AL   
15  Limited Liability company Yes       AL   
16  Corporation               Yes       AL   
17  Corporation               Yes       AL   
18  Corporation               Yes       AL   
19  Corporation               Yes       AL   
20  Other                     No        AL




Number of For Profit and and Non Profit Nursing Homes in MN.



ForProfit Count
1 No        253  
2 Yes       114

Next, let us consider the various levels for Corporation Type.

In [68]:
(
 NursingHomes_OwnershipSplit
  %>% select(CorporationType)
  %>% distinct(CorporationType)   
 ) -> temp

 head(temp, 15)

CorporationType           
1   Corporation              
2   County                   
3   Individual               
4   Limited Liability company
5   Other                    
6   Partnership              
7   City                     
8   Church related           
9   City/county              
10  Federal                  
11  State                    
12  Hospital district

Consider the following recoding regarding the Corporation Type.

<table>
<tr>
<td align="center">CorporationType</td>
<td align="center">Corporation</td>
</tr>
<tr>
 <td align="center"><font size="+1">Corporation</font></td>
 <td align="center"><font size="+1">Yes</font></td>
</tr>
<tr>
 <td align="center"><font size="+1">Limited Liability company</font></td>
  <td align="center"><font size="+1">Yes</font></td>
</tr>

<tr>
 <td align="center"><font size="+1">County</font></td>
 <td align="center" rowspan="10"><font size="+1">No</font></td>
</tr>
<tr>
 <td align="center"><font size="+1">Individual</font></td>
 
</tr>

<tr>
 <td align="center"><font size="+1">Other</font></td>

</tr>
<tr>
 <td align="center"><font size="+1">Partnership</font></td>

</tr>
<tr>
 <td align="center"><font size="+1">City</font></td>

</tr>
<tr>
 <td align="center"><font size="+1">Church related</font></td>
 
</tr>
<tr>
 <td align="center"><font size="+1">City/county</font></td>

</tr>
<tr>
 <td align="center"><font size="+1">Federal</font></td>
 
</tr>
<tr>
 <td align="center"><font size="+1">State</font></td>
 
</tr>
<tr>
 <td align="center"><font size="+1">Hospital district</font></td>
 
</tr>
</table>

A couple of comments regarding the use of the <strong>case_when()</strong> function.


1.   The <strong>order</strong> of the logical statements matter, e.g. statements are considered from top-down.
2.   If it is <strong>TRUE</strong> that <strong>none</strong> of the other conditions apply, the TRUE can be placed on the right-hand side of the ~ to consider all remaining situations.



In [73]:
(
 NursingHomes_OwnershipSplit
  %>% mutate(Corporation = 
                          case_when(
                                    str_trim(CorporationType) == "Corporation" ~ "Yes",
                                    str_trim(CorporationType) == "Limited Liability company" ~ "Yes",
                                    TRUE ~ "No"
                          )
  
            )
   %>% select(OwnershipType_Split, ProfitType, CorporationType, Corporation, State)
 ) -> temp

 head(temp,20)

 #The Corporation variable can now be used, e.g. count the number of Corporation type nursing homes in MN
cat("\n")
cat("\n")
cat("\n")
cat("Number of Corporation vs Other types Nursing Homes in MN.\n")
cat("\n")
(
 temp
 %>% filter(State == "MN")
 %>% group_by(Corporation)
 %>% summarize(Count=n())
)

OwnershipType_Split                                    ProfitType 
1  For profit  ,  Corporation                             For profit 
2  For profit  ,  Corporation                             For profit 
3  Government ,  County                                   Government 
4  For profit ,  Individual                               For profit 
5  For profit ,  Individual                               For profit 
6  For profit  ,  Corporation                             For profit 
7  For profit  ,  Corporation                             For profit 
8  For profit  ,  Corporation                             For profit 
9  For profit  ,  Corporation                             For profit 
10 For profit  ,  Corporation                             For profit 
11 For profit  ,  Corporation                             For profit 
12 For profit  ,  Corporation                             For profit 
13 For profit  ,  Corporation                             For profit 
14 For profit  ,  Corporation                             For profit 
15 For profit                ,  Limited Liability company For profit 
16 For profit  ,  Corporation                             For profit 
17 For profit  ,  Corporation                             For profit 
18 For profit  ,  Corporation                             For profit 
19 For profit  ,  Corporation                             For profit 
20 Non profit ,  Other                                    Non profit 
   CorporationType            Corporation State
1   Corporation               Yes         AL   
2   Corporation               Yes         AL   
3   County                    No          AL   
4   Individual                No          AL   
5   Individual                No          AL   
6   Corporation               Yes         AL   
7   Corporation               Yes         AL   
8   Corporation               Yes         AL   
9   Corporation               Yes         AL   
10  Corporation               Yes         AL   
11  Corporation               Yes         AL   
12  Corporation               Yes         AL   
13  Corporation               Yes         AL   
14  Corporation               Yes         AL   
15  Limited Liability company Yes         AL   
16  Corporation               Yes         AL   
17  Corporation               Yes         AL   
18  Corporation               Yes         AL   
19  Corporation               Yes         AL   
20  Other                     No          AL




Number of Corporation vs Other types Nursing Homes in MN.



Corporation Count
1 No          100  
2 Yes         267



---

---







**Task #3**:  Use the str_split() function to retrieve the State from the ProviderAddress field. 

In [32]:
(
 NursingHomes
 %>% select(ProviderAddress)
 %>% head()
)

ProviderAddress                              
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653   
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150    
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768      
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206 
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111
6 611 WEST MARKET STREET,ATHENS,AL,35611